compiler/il/ILProps.hpp

```diff
@@ -47,6 +49,107 @@ namespace ILTypeProp
+ * 0x02_00_00_00_00_00_00_00 | (TR::Int16 << 8) | TR::Int8
+ *    means a node expects two children: the first an 8-bit integer,
+ *    the second a 16-bit integer
+ *
+ * 0x01_00_00_00_00_00_00_ff
+ *    means a node expects one child of an unspecified type
+ *
+ * 0xff_ff_ff_ff_ff_ff_ff_ff
+ *    means a node expects an unspecified number of children of unspecified type
+ *
+ * 0xff_ff_ff_ff_ff_ff_00_00 | (TR::Int64 << 8) | TR::Int32
+ *    means a node expects an unspecified number of children with the first
+ *    child being a 32-bit integer, the second child being a 64-bit integer, and
+ *    the type of any other chidlren is unspecified
+ *
+ * To simplify handling of these values, some macros are defined for correctly
+ * composing and decomposing them using bit-masks. Some default values are also
+ * provided as static const variables. Although this could be implemented as an
+ * enum, some compilers used for testing had problems handling 64-bit enums.
+ */
+namespace TR { typedef uint64_t ILChildPropType; }
+namespace ILChildProp
+   {
+
+   static const TR::ILChildPropType NoChildren = 0;
+   static const TR::ILChildPropType UnspecifiedChildType = 0xff;
+   static const TR::ILChildPropType UnspecifiedChildCount = 0xff;
+   static const TR::ILChildPropType Unspecified = 0xffffffffffffffff;
+
+   static const TR::ILChildPropType IndirectCallType = 0xffffffffffffff00 | TR::Address;
+
+   static const TR::ILChildPropType FirstChildTypeMask   =               0xff;
+   static const TR::ILChildPropType SecondChildTypeMask  =             0xff00;
+   static const TR::ILChildPropType ThirdChildTypeMask   =           0xff0000;
+   static const TR::ILChildPropType FourthChildTypeMask  =         0xff000000;
+   static const TR::ILChildPropType FifthChildTypeMask   =       0xff00000000;
+   static const TR::ILChildPropType SixthChildTypeMask   =     0xff0000000000;
+   static const TR::ILChildPropType SeventhChildTypeMask =   0xff000000000000;
+   static const TR::ILChildPropType ChildCountMask       = 0xff00000000000000;
+   }
+
+// make sure that all TR::DataTypes can fit in one byte
+// (remembering that 255 is a reserved value)
+static_assert(TR::NumTypes < 254, "There are too many data types to fit in one byte.");
+
+#define FIRST_CHILD(type)     (type) <<  0
+#define SECOND_CHILD(type)    (type) <<  8
+#define THIRD_CHILD(type)     (type) << 16
+#define FOURTH_CHILD(type)    (type) << 24
+#define FIFTH_CHILD(type)     (type) << 32
+#define SIXTH_CHILD(type)     (type) << 40
+#define SEVENTH_CHILD(type)   (type) << 48
+#define CHILD_COUNT(c)        (c)    << 56
+
+#define ONE_CHILD(type)                                              CHILD_COUNT(1L) | FIRST_CHILD(type)
+#define TWO_CHILD(type1, type2)                                      CHILD_COUNT(2L) | FIRST_CHILD(type1) | SECOND_CHILD(type2)
+#define THREE_CHILD(type1, type2, type3)                             CHILD_COUNT(3L) | FIRST_CHILD(type1) | SECOND_CHILD(type2) | THIRD_CHILD(type3)
```

Comment:
Why aren't `FOUR_CHILD` and `SIX_CHILD` defined?  Even if there may be no opcodes that need this many children I think its good to provide them for completeness.